File: auth_client.py
"""
Microsoft Authentication Client for Fabric Analytics
Supports Bearer Token, Service Principal, and Device Code authentication methods
"""

import json
import time
import webbrowser
from datetime import datetime, timedelta
from typing import Dict, Optional, Tuple, Union
from dataclasses import dataclass
from enum import Enum
import sys

try:
    import msal
    MSAL_AVAILABLE = True
except ImportError:
    MSAL_AVAILABLE = False

class AuthMethod(Enum):
    """Authentication methods supported by the client"""
    BEARER_TOKEN = "bearer"
    SERVICE_PRINCIPAL = "service_principal"
    DEVICE_CODE = "device_code"
    INTERACTIVE = "interactive"

@dataclass
class AuthResult:
    """Result of authentication containing access token and metadata"""
    access_token: str
    expires_on: datetime
    account: Optional[Dict] = None
    refresh_token: Optional[str] = None
    
    def is_valid(self) -> bool:
        """Check if the token is still valid"""
        return self.expires_on > datetime.now()
    
    def expires_in_seconds(self) -> int:
        """Get seconds until token expires"""
        delta = self.expires_on - datetime.now()
        return max(0, int(delta.total_seconds()))

class MicrosoftAuthClient:
    """Microsoft Authentication Client supporting multiple auth flows"""
    
    DEFAULT_SCOPE = ["https://api.fabric.microsoft.com/.default"]
    DEFAULT_AUTHORITY = "https://login.microsoftonline.com/common"
    
    def __init__(self, client_id: Optional[str] = None, authority: Optional[str] = None):
        """
        Initialize the authentication client
        
        Args:
            client_id: Application (client) ID
            authority: Authority URL (defaults to common endpoint)
        """
        self.client_id = client_id
        self.authority = authority or self.DEFAULT_AUTHORITY
        self.app: Optional[msal.ClientApplication] = None
        self._check_msal_availability()
    
    def _check_msal_availability(self):
        """Check if MSAL is available and provide installation instructions if not"""
        if not MSAL_AVAILABLE:
            print("\n❌ MSAL (Microsoft Authentication Library) is not installed.")
            print("To use Microsoft authentication, please install it:")
            print("pip install msal")
            print("\nFalling back to bearer token authentication only.")
    
    def authenticate_with_bearer_token(self, token: str) -> AuthResult:
        """
        Authenticate using a bearer token
        
        Args:
            token: Bearer token (with or without 'Bearer ' prefix)
            
        Returns:
            AuthResult with the provided token
        """
        # Clean up token format
        if token.startswith('Bearer '):
            access_token = token[7:]  # Remove 'Bearer ' prefix
        else:
            access_token = token
        
        # Basic token validation (check if it's JWT-like)
        parts = access_token.split('.')
        if len(parts) != 3:
            raise ValueError("Invalid bearer token format. Token does not appear to be a valid JWT.")
        
        # Default expiry to 1 hour from now
        expires_on = datetime.now() + timedelta(hours=1)
        
        return AuthResult(
            access_token=access_token,
            expires_on=expires_on
        )
    
    def authenticate_with_service_principal(
        self, 
        client_id: str, 
        client_secret: str, 
        tenant_id: str,
        scope: Optional[list] = None
    ) -> AuthResult:
        """
        Authenticate using service principal (client credentials flow)
        
        Args:
            client_id: Application (client) ID
            client_secret: Client secret
            tenant_id: Tenant ID
            scope: List of scopes (defaults to Fabric default scope)
            
        Returns:
            AuthResult with access token
        """
        if not MSAL_AVAILABLE:
            raise RuntimeError("MSAL is required for service principal authentication. Please install with: pip install msal")
        
        authority = f"https://login.microsoftonline.com/{tenant_id}"
        scope = scope or self.DEFAULT_SCOPE
        
        # Create confidential client application
        app = msal.ConfidentialClientApplication(
            client_id=client_id,
            client_credential=client_secret,
            authority=authority
        )
        
        # Acquire token
        result = app.acquire_token_for_client(scopes=scope)
        
        if "error" in result:
            error_msg = result.get("error_description", result.get("error", "Unknown error"))
            raise RuntimeError(f"Service principal authentication failed: {error_msg}")
        
        # Calculate expiry time
        expires_in = result.get("expires_in", 3600)  # Default to 1 hour
        expires_on = datetime.now() + timedelta(seconds=expires_in)
        
        return AuthResult(
            access_token=result["access_token"],
            expires_on=expires_on,
            account=result.get("account")
        )
    
    def authenticate_with_device_code(
        self, 
        client_id: str, 
        tenant_id: Optional[str] = None,
        scope: Optional[list] = None
    ) -> AuthResult:
        """
        Authenticate using device code flow
        
        Args:
            client_id: Application (client) ID
            tenant_id: Tenant ID (optional, defaults to common)
            scope: List of scopes (defaults to Fabric default scope)
            
        Returns:
            AuthResult with access token
        """
        if not MSAL_AVAILABLE:
            raise RuntimeError("MSAL is required for device code authentication. Please install with: pip install msal")
        
        authority = f"https://login.microsoftonline.com/{tenant_id}" if tenant_id else self.authority
        scope = scope or self.DEFAULT_SCOPE
        
        # Create public client application
        app = msal.PublicClientApplication(
            client_id=client_id,
            authority=authority
        )
        
        # Initiate device flow
        flow = app.initiate_device_flow(scopes=scope)
        
        if "user_code" not in flow:
            raise RuntimeError("Failed to create device flow")
        
        # Display device code to user
        print("\n" + "="*50)
        print("🔐 DEVICE CODE AUTHENTICATION")
        print("="*50)
        print(f"📱 Please visit: {flow['verification_uri']}")
        print(f"🔑 Enter the code: {flow['user_code']}")
        print(f"⏰ Code expires in: {flow['expires_in']} seconds")
        print("⏳ Waiting for authentication...")
        print("="*50 + "\n")
        
        # Complete device flow
        result = app.acquire_token_by_device_flow(flow)
        
        if "error" in result:
            error_msg = result.get("error_description", result.get("error", "Unknown error"))
            raise RuntimeError(f"Device code authentication failed: {error_msg}")
        
        # Calculate expiry time
        expires_in = result.get("expires_in", 3600)  # Default to 1 hour
        expires_on = datetime.now() + timedelta(seconds=expires_in)
        
        print("✅ Authentication successful!")
        
        return AuthResult(
            access_token=result["access_token"],
            expires_on=expires_on,
            account=result.get("account"),
            refresh_token=result.get("refresh_token")
        )
    
    def authenticate_interactively(
        self, 
        client_id: str, 
        tenant_id: Optional[str] = None,
        scope: Optional[list] = None
    ) -> AuthResult:
        """
        Authenticate using interactive flow (opens browser)
        
        Args:
            client_id: Application (client) ID
            tenant_id: Tenant ID (optional, defaults to common)
            scope: List of scopes (defaults to Fabric default scope)
            
        Returns:
            AuthResult with access token
        """
        if not MSAL_AVAILABLE:
            raise RuntimeError("MSAL is required for interactive authentication. Please install with: pip install msal")
        
        authority = f"https://login.microsoftonline.com/{tenant_id}" if tenant_id else self.authority
        scope = scope or self.DEFAULT_SCOPE
        
        # Create public client application
        app = msal.PublicClientApplication(
            client_id=client_id,
            authority=authority
        )
        
        print("\n" + "="*50)
        print("🌐 INTERACTIVE AUTHENTICATION")
        print("="*50)
        print("🔄 Opening browser for authentication...")
        print("="*50 + "\n")
        
        # Try to acquire token interactively
        result = app.acquire_token_interactive(scopes=scope)
        
        if "error" in result:
            error_msg = result.get("error_description", result.get("error", "Unknown error"))
            raise RuntimeError(f"Interactive authentication failed: {error_msg}")
        
        # Calculate expiry time
        expires_in = result.get("expires_in", 3600)  # Default to 1 hour
        expires_on = datetime.now() + timedelta(seconds=expires_in)
        
        print("✅ Authentication successful!")
        
        return AuthResult(
            access_token=result["access_token"],
            expires_on=expires_on,
            account=result.get("account"),
            refresh_token=result.get("refresh_token")
        )
    
    @staticmethod
    def prompt_for_auth_method() -> AuthMethod:
        """
        Prompt user to select authentication method
        
        Returns:
            Selected AuthMethod
        """
        print("\n" + "="*60)
        print("🔐 AUTHENTICATION METHOD SELECTION")
        print("="*60)
        print("1. 🎫 Bearer Token (provide your own token)")
        print("2. 🤖 Service Principal (client ID + secret + tenant)")
        print("3. 📱 Device Code (sign in with browser on another device)")
        print("4. 🌐 Interactive (sign in with browser - opens automatically)")
        print("="*60)
        
        while True:
            try:
                choice = input("Enter your choice (1-4): ").strip()
                if choice == "1":
                    return AuthMethod.BEARER_TOKEN
                elif choice == "2":
                    return AuthMethod.SERVICE_PRINCIPAL
                elif choice == "3":
                    return AuthMethod.DEVICE_CODE
                elif choice == "4":
                    return AuthMethod.INTERACTIVE
                else:
                    print("❌ Invalid choice. Please enter 1, 2, 3, or 4.")
            except KeyboardInterrupt:
                print("\n🚫 Authentication cancelled by user.")
                sys.exit(0)
    
    @staticmethod
    def prompt_for_bearer_token() -> str:
        """
        Prompt user to enter bearer token
        
        Returns:
            Bearer token string
        """
        print("\n" + "="*60)
        print("🎫 BEARER TOKEN INPUT")
        print("="*60)
        print("Please enter your bearer token:")
        print("• Include 'Bearer ' prefix if you have it")
        print("• Or just paste the token value")
        print("="*60)
        
        token = input("Bearer token: ").strip()
        if not token:
            raise ValueError("Bearer token cannot be empty")
        
        return token
    
    @staticmethod
    def prompt_for_service_principal() -> Tuple[str, str, str]:
        """
        Prompt user to enter service principal credentials
        
        Returns:
            Tuple of (client_id, client_secret, tenant_id)
        """
        print("\n" + "="*60)
        print("🤖 SERVICE PRINCIPAL CREDENTIALS")
        print("="*60)
        print("Please enter your service principal details:")
        print("="*60)
        
        client_id = input("Client ID (Application ID): ").strip()
        if not client_id:
            raise ValueError("Client ID cannot be empty")
        
        client_secret = input("Client Secret: ").strip()
        if not client_secret:
            raise ValueError("Client Secret cannot be empty")
        
        tenant_id = input("Tenant ID (Directory ID): ").strip()
        if not tenant_id:
            raise ValueError("Tenant ID cannot be empty")
        
        return client_id, client_secret, tenant_id
    
    @staticmethod
    def prompt_for_device_code_details() -> Tuple[str, Optional[str]]:
        """
        Prompt user to enter device code authentication details
        
        Returns:
            Tuple of (client_id, tenant_id)
        """
        print("\n" + "="*60)
        print("📱 DEVICE CODE AUTHENTICATION")
        print("="*60)
        print("Please enter your application details:")
        print("="*60)
        
        client_id = input("Client ID (Application ID): ").strip()
        if not client_id:
            raise ValueError("Client ID cannot be empty")
        
        tenant_id = input("Tenant ID (optional, press Enter for common): ").strip()
        if not tenant_id:
            tenant_id = None
        
        return client_id, tenant_id
    
    @staticmethod
    def prompt_for_interactive_details() -> Tuple[str, Optional[str]]:
        """
        Prompt user to enter interactive authentication details
        
        Returns:
            Tuple of (client_id, tenant_id)
        """
        print("\n" + "="*60)
        print("🌐 INTERACTIVE AUTHENTICATION")
        print("="*60)
        print("Please enter your application details:")
        print("="*60)
        
        client_id = input("Client ID (Application ID): ").strip()
        if not client_id:
            raise ValueError("Client ID cannot be empty")
        
        tenant_id = input("Tenant ID (optional, press Enter for common): ").strip()
        if not tenant_id:
            tenant_id = None
        
        return client_id, tenant_id
    
    def authenticate(self) -> AuthResult:
        """
        Interactive authentication flow - prompts user for method and credentials
        
        Returns:
            AuthResult with access token
        """
        try:
            auth_method = self.prompt_for_auth_method()
            
            if auth_method == AuthMethod.BEARER_TOKEN:
                token = self.prompt_for_bearer_token()
                return self.authenticate_with_bearer_token(token)
            
            elif auth_method == AuthMethod.SERVICE_PRINCIPAL:
                client_id, client_secret, tenant_id = self.prompt_for_service_principal()
                return self.authenticate_with_service_principal(client_id, client_secret, tenant_id)
            
            elif auth_method == AuthMethod.DEVICE_CODE:
                client_id, tenant_id = self.prompt_for_device_code_details()
                return self.authenticate_with_device_code(client_id, tenant_id)
            
            elif auth_method == AuthMethod.INTERACTIVE:
                client_id, tenant_id = self.prompt_for_interactive_details()
                return self.authenticate_interactively(client_id, tenant_id)
            
            else:
                raise ValueError(f"Unsupported authentication method: {auth_method}")
                
        except KeyboardInterrupt:
            print("\n🚫 Authentication cancelled by user.")
            sys.exit(0)
        except Exception as e:
            print(f"\n❌ Authentication failed: {e}")
            raise

def main():
    """Demo/test function for the authentication client"""
    print("🔐 Microsoft Fabric Authentication Client Demo")
    print("=" * 60)
    
    client = MicrosoftAuthClient()
    
    try:
        auth_result = client.authenticate()
        
        print("\n✅ Authentication successful!")
        print(f"📅 Token expires: {auth_result.expires_on}")
        print(f"⏱️  Expires in: {auth_result.expires_in_seconds()} seconds")
        
        if auth_result.account:
            print(f"👤 Account: {auth_result.account.get('username', 'Unknown')}")
        
        # Test token format
        print(f"\n🔑 Token preview: {auth_result.access_token[:20]}...")
        
        return auth_result
        
    except Exception as e:
        print(f"\n❌ Authentication failed: {e}")
        return None

if __name__ == "__main__":
    main()

File: src/auth-client.ts
import { 
  PublicClientApplication, 
  ConfidentialClientApplication,
  DeviceCodeRequest,
  ClientCredentialRequest,
  SilentFlowRequest,
  AccountInfo,
  Configuration,
  LogLevel
} from "@azure/msal-node";
import { createServer } from "http";
import { URL } from "url";
import { exec } from "child_process";
import { promisify } from "util";

const execAsync = promisify(exec);

export interface AuthConfig {
  clientId: string;
  authority?: string;
  redirectUri?: string;
  clientSecret?: string;
  scope?: string[];
}

export interface AuthResult {
  accessToken: string;
  expiresOn: Date;
  account?: AccountInfo;
}

/* eslint-disable no-unused-vars */
export enum AuthMethod {
  BEARER_TOKEN = "bearer",
  SERVICE_PRINCIPAL = "service_principal", 
  DEVICE_CODE = "device_code",
  INTERACTIVE = "interactive",
  AZURE_CLI = "azure_cli"
}
/* eslint-enable no-unused-vars */

export class MicrosoftAuthClient {
  private publicClient?: PublicClientApplication;
  private confidentialClient?: ConfidentialClientApplication;
  private config: AuthConfig;
  private readonly defaultScope = ["https://api.fabric.microsoft.com/.default"];

  constructor(config: AuthConfig) {
    this.config = {
      ...config,
      authority: config.authority || "https://login.microsoftonline.com/common",
      scope: config.scope || this.defaultScope
    };
  }

  /**
   * Initialize the appropriate MSAL client based on auth method
   */
  private initializeClient(method: AuthMethod): void {
    const clientConfig: Configuration = {
      auth: {
        clientId: this.config.clientId,
        authority: this.config.authority,
      },
      system: {
        loggerOptions: {
          loggerCallback: (level, message, _containsPii) => {
            if (level === LogLevel.Error) {
              console.error("MSAL Error:", message);
            }
          },
          piiLoggingEnabled: false,
          logLevel: LogLevel.Warning,
        },
      },
    };

    if (method === AuthMethod.SERVICE_PRINCIPAL && this.config.clientSecret) {
      // Use confidential client for service principal
      this.confidentialClient = new ConfidentialClientApplication({
        ...clientConfig,
        auth: {
          ...clientConfig.auth,
          clientSecret: this.config.clientSecret,
        },
      });
    } else {
      // Use public client for device code and interactive flows
      this.publicClient = new PublicClientApplication(clientConfig);
    }
  }

  /**
   * Authenticate using bearer token (direct token provision)
   */
  async authenticateWithBearerToken(token: string): Promise<AuthResult> {
    // For bearer token, we don't need MSAL - just validate and return
    if (!token || !token.startsWith('Bearer ')) {
      throw new Error("Invalid bearer token format. Expected format: 'Bearer <token>'");
    }

    const accessToken = token.substring(7); // Remove 'Bearer ' prefix
    
    // Basic token validation (check if it's a JWT-like structure)
    const parts = accessToken.split('.');
    if (parts.length !== 3) {
      throw new Error("Invalid bearer token format. Token does not appear to be a valid JWT.");
    }

    return {
      accessToken,
      expiresOn: new Date(Date.now() + 3600000), // Default 1 hour expiry
    };
  }

  /**
   * Authenticate using service principal (client credentials flow)
   */
  async authenticateWithServicePrincipal(
    clientId: string,
    clientSecret: string,
    tenantId: string
  ): Promise<AuthResult> {
    this.config.clientId = clientId;
    this.config.clientSecret = clientSecret;
    this.config.authority = `https://login.microsoftonline.com/${tenantId}`;
    
    this.initializeClient(AuthMethod.SERVICE_PRINCIPAL);

    if (!this.confidentialClient) {
      throw new Error("Failed to initialize confidential client for service principal authentication");
    }

    const clientCredentialRequest: ClientCredentialRequest = {
      scopes: this.config.scope!,
      skipCache: false,
    };

    try {
      const response = await this.confidentialClient.acquireTokenByClientCredential(clientCredentialRequest);
      
      if (!response || !response.accessToken) {
        throw new Error("Failed to acquire token using service principal");
      }

      return {
        accessToken: response.accessToken,
        expiresOn: response.expiresOn || new Date(Date.now() + 3600000),
        account: response.account || undefined,
      };
    } catch (error) {
      throw new Error(`Service principal authentication failed: ${error instanceof Error ? error.message : String(error)}`);
    }
  }

  /**
   * Authenticate using device code flow (for headless environments)
   */
  async authenticateWithDeviceCode(clientId: string, tenantId?: string): Promise<AuthResult> {
    this.config.clientId = clientId;
    if (tenantId) {
      this.config.authority = `https://login.microsoftonline.com/${tenantId}`;
    }
    
    this.initializeClient(AuthMethod.DEVICE_CODE);

    if (!this.publicClient) {
      throw new Error("Failed to initialize public client for device code authentication");
    }

    const deviceCodeRequest: DeviceCodeRequest = {
      scopes: this.config.scope!,
      deviceCodeCallback: (response) => {
        console.log("\n=== DEVICE CODE AUTHENTICATION ===");
        console.log(`Please visit: ${response.verificationUri}`);
        console.log(`And enter the code: ${response.userCode}`);
        console.log(`Expires in: ${response.expiresIn} seconds`);
        console.log("Waiting for authentication...\n");
      },
    };

    try {
      const response = await this.publicClient.acquireTokenByDeviceCode(deviceCodeRequest);
      
      if (!response || !response.accessToken) {
        throw new Error("Failed to acquire token using device code");
      }

      return {
        accessToken: response.accessToken,
        expiresOn: response.expiresOn || new Date(Date.now() + 3600000),
        account: response.account || undefined,
      };
    } catch (error) {
      throw new Error(`Device code authentication failed: ${error instanceof Error ? error.message : String(error)}`);
    }
  }

  /**
   * Authenticate using interactive flow (opens browser)
   */
  async authenticateInteractively(clientId: string, tenantId?: string): Promise<AuthResult> {
    this.config.clientId = clientId;
    this.config.redirectUri = "http://localhost:8080";
    if (tenantId) {
      this.config.authority = `https://login.microsoftonline.com/${tenantId}`;
    }
    
    this.initializeClient(AuthMethod.INTERACTIVE);

    if (!this.publicClient) {
      throw new Error("Failed to initialize public client for interactive authentication");
    }

    return new Promise((resolve, reject) => {
      const server = createServer((req, res) => {
        if (req.url && req.url.startsWith('/?code=')) {
          const url = new URL(req.url, 'http://localhost:8080');
          const code = url.searchParams.get('code');
          
          res.writeHead(200, { 'Content-Type': 'text/html' });
          res.end(`
            <html>
              <body>
                <h2>Authentication successful!</h2>
                <p>You can close this window and return to the application.</p>
              </body>
            </html>
          `);
          
          server.close();
          
          if (code) {
            this.handleAuthorizationCode(code).then(resolve).catch(reject);
          } else {
            reject(new Error("No authorization code received"));
          }
        }
      });

      server.listen(8080, () => {
        const authUrl = this.buildAuthUrl();
        console.log("\n=== INTERACTIVE AUTHENTICATION ===");
        console.log(`Please visit: ${authUrl}`);
        console.log("Waiting for authentication...\n");
        
        // Auto-open browser if possible
        import('child_process').then(({ exec }) => {
          exec(`start ${authUrl}`, (error) => {
            if (error) {
              console.log("Could not auto-open browser. Please manually visit the URL above.");
            }
          });
        });
      });

      // Timeout after 5 minutes
      setTimeout(() => {
        server.close();
        reject(new Error("Authentication timeout"));
      }, 300000);
    });
  }

  private buildAuthUrl(): string {
    const params = new URLSearchParams({
      client_id: this.config.clientId,
      response_type: 'code',
      redirect_uri: this.config.redirectUri!,
      scope: this.config.scope!.join(' '),
      response_mode: 'query',
    });

    return `${this.config.authority}/oauth2/v2.0/authorize?${params.toString()}`;
  }

  private async handleAuthorizationCode(_code: string): Promise<AuthResult> {
    // This is a simplified implementation - in a real scenario, you'd exchange the code for tokens
    // For now, we'll use device code flow as a fallback
    console.log("Received authorization code, exchanging for tokens...");
    
    // Fallback to device code flow
    return this.authenticateWithDeviceCode(this.config.clientId);
  }

  /**
   * Get token silently (from cache)
   */
  async getTokenSilently(account: AccountInfo): Promise<AuthResult | null> {
    if (!this.publicClient) {
      return null;
    }

    const silentRequest: SilentFlowRequest = {
      scopes: this.config.scope!,
      account: account,
    };

    try {
      const response = await this.publicClient.acquireTokenSilent(silentRequest);
      
      if (!response || !response.accessToken) {
        return null;
      }

      return {
        accessToken: response.accessToken,
        expiresOn: response.expiresOn || new Date(Date.now() + 3600000),
        account: response.account || undefined,
      };
    // eslint-disable-next-line @typescript-eslint/no-unused-vars, no-unused-vars
    } catch (_error) {
      return null; // Silent acquisition failed, need interactive flow
    }
  }

  /**
   * Get all cached accounts
   */
  async getCachedAccounts(): Promise<AccountInfo[]> {
    if (!this.publicClient) {
      return [];
    }

    return this.publicClient.getTokenCache().getAllAccounts();
  }

  /**
   * Clear token cache
   */
  async clearCache(): Promise<void> {
    if (this.publicClient) {
      const accounts = await this.getCachedAccounts();
      for (const account of accounts) {
        await this.publicClient.getTokenCache().removeAccount(account);
      }
    }
  }

  /**
   * Validate if token is still valid
   */
  isTokenValid(authResult: AuthResult): boolean {
    return authResult.expiresOn > new Date();
  }

  /**
   * Get authentication method prompt for CLI
   */
  static getAuthMethodPrompt(): string {
    return `
Choose authentication method:
1. Bearer Token (provide your own token)
2. Service Principal (client ID + secret + tenant)
3. Device Code (sign in with browser on another device)
4. Interactive (sign in with browser - opens automatically)
5. Azure CLI (use Azure CLI logged-in context)

Enter choice (1-5): `;
  }

  /**
   * Parse authentication method from user input
   */
  static parseAuthMethod(input: string): AuthMethod {
    switch (input.trim()) {
      case "1":
        return AuthMethod.BEARER_TOKEN;
      case "2":
        return AuthMethod.SERVICE_PRINCIPAL;
      case "3":
        return AuthMethod.DEVICE_CODE;
      case "4":
        return AuthMethod.INTERACTIVE;
      case "5":
        return AuthMethod.AZURE_CLI;
      default:
        throw new Error("Invalid authentication method selection");
    }
  }

  /**
   * Authenticate using Azure CLI context
   * This method leverages the existing Azure CLI login session
   */
  async authenticateWithAzureCli(scope?: string[]): Promise<AuthResult> {
    try {
      // Check if Azure CLI is installed and user is logged in
      await this.checkAzureCliStatus();
      
      // Get access token using Azure CLI
      const targetScope = scope || this.config.scope || this.defaultScope;
      const scopeString = targetScope.join(' ');
      
      // Use Azure CLI to get token for Microsoft Fabric/Power BI
      const command = `az account get-access-token --scope "${scopeString}" --output json`;
      
      console.log("🔄 Getting access token from Azure CLI...");
      console.log(`   Scope: ${scopeString}`);
      
      const { stdout } = await execAsync(command);
      const tokenResponse = JSON.parse(stdout);
      
      if (!tokenResponse.accessToken) {
        throw new Error("Failed to get access token from Azure CLI - no access token in response");
      }
      
      // Parse expiration date
      const expiresOn = new Date(tokenResponse.expiresOn || Date.now() + 3600000);
      
      console.log("✅ Successfully authenticated using Azure CLI");
      console.log(`   Token expires: ${expiresOn.toISOString()}`);
      console.log(`   Token length: ${tokenResponse.accessToken.length} characters`);
      
      return {
        accessToken: tokenResponse.accessToken,
        expiresOn: expiresOn,
        account: undefined // Azure CLI doesn't provide account info in the same format
      };
      
    } catch (error) {
      console.error("❌ Azure CLI authentication failed:", error instanceof Error ? error.message : String(error));
      console.error("🔧 Troubleshooting steps:");
      console.error("   1. Ensure Azure CLI is installed: az --version");
      console.error("   2. Ensure you're logged in: az account show");
      console.error("   3. Try refreshing login: az account get-access-token --scope https://api.fabric.microsoft.com/.default");
      throw new Error(`Azure CLI authentication failed: ${error instanceof Error ? error.message : String(error)}`);
    }
  }

  /**
   * Check Azure CLI installation and login status
   */
  private async checkAzureCliStatus(): Promise<void> {
    try {
      // Check if Azure CLI is installed
      await execAsync('az --version');
      console.log("✅ Azure CLI is installed");
      
      // Check if user is logged in
      const { stdout } = await execAsync('az account show --output json');
      const account = JSON.parse(stdout);
      
      if (!account || !account.id) {
        throw new Error("No Azure account found. Please run 'az login' first.");
      }
      
      console.log(`✅ Logged in as: ${account.user?.name || account.user?.type || 'Unknown'}`);
      console.log(`   Subscription: ${account.name} (${account.id})`);
      console.log(`   Tenant: ${account.tenantId}`);
      
      // Test access to Microsoft Fabric specifically
      try {
        await execAsync('az account get-access-token --scope "https://api.fabric.microsoft.com/.default" --query "accessToken" --output tsv');
        console.log("✅ Microsoft Fabric API access confirmed");
      } catch (fabricError) {
        console.warn("⚠️  Warning: Could not verify Fabric API access, but authentication may still work");
        console.warn(`    Error: ${fabricError instanceof Error ? fabricError.message : String(fabricError)}`);
      }
      
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      
      if (errorMessage.includes('az: command not found') || errorMessage.includes("'az' is not recognized")) {
        throw new Error(`
Azure CLI is not installed or not in PATH. 
Please install Azure CLI from: https://docs.microsoft.com/en-us/cli/azure/install-azure-cli
Then run 'az login' to authenticate.`);
      }
      
      if (errorMessage.includes('Please run')) {
        throw new Error(`
Azure CLI is installed but you are not logged in.
Please run 'az login' to authenticate with your Azure account.`);
      }
      
      throw error;
    }
  }

  /**
   * Validate that Azure CLI authentication is properly configured for MCP usage
   * This is useful for troubleshooting Claude Desktop integration
   */
  async validateAzureCliForMCP(): Promise<{valid: boolean, message: string, recommendations?: string[]}> {
    try {
      console.log("🔍 Validating Azure CLI configuration for MCP usage...");
      
      // Step 1: Check Azure CLI installation
      await execAsync('az --version');
      
      // Step 2: Check login status
      const { stdout: accountInfo } = await execAsync('az account show --output json');
      const account = JSON.parse(accountInfo);
      
      if (!account || !account.id) {
        return {
          valid: false,
          message: "Azure CLI not logged in",
          recommendations: [
            "Run 'az login' to authenticate",
            "Ensure you have access to Microsoft Fabric in your tenant"
          ]
        };
      }
      
      // Step 3: Test Fabric token acquisition
      const { stdout: tokenInfo } = await execAsync('az account get-access-token --scope "https://api.fabric.microsoft.com/.default" --output json');
      const tokenResponse = JSON.parse(tokenInfo);
      
      if (!tokenResponse.accessToken) {
        return {
          valid: false,
          message: "Cannot acquire Fabric access token",
          recommendations: [
            "Verify you have Microsoft Fabric permissions",
            "Check if your tenant has Fabric enabled",
            "Try running: az account get-access-token --scope https://api.fabric.microsoft.com/.default"
          ]
        };
      }
      
      // Step 4: Test a simple Fabric API call
      try {
        await execAsync(`az rest --method GET --url "https://api.fabric.microsoft.com/v1/workspaces" --resource "https://api.fabric.microsoft.com/" --query "value[0:1]"`);
        
        return {
          valid: true,
          message: `Azure CLI authentication configured correctly for user: ${account.user?.name || 'Unknown'}`
        };
      } catch (_apiError) {
        return {
          valid: false,
          message: "Token acquired but Fabric API call failed",
          recommendations: [
            "Verify you have permissions to access Microsoft Fabric",
            "Check if your organization allows Fabric API access",
            "Contact your administrator about Fabric access permissions"
          ]
        };
      }
      
    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      
      if (errorMessage.includes('az: command not found') || errorMessage.includes("'az' is not recognized")) {
        return {
          valid: false,
          message: "Azure CLI not installed",
          recommendations: [
            "Install Azure CLI from: https://docs.microsoft.com/en-us/cli/azure/install-azure-cli",
            "Add Azure CLI to your system PATH",
            "Restart your terminal/VS Code after installation"
          ]
        };
      }
      
      return {
        valid: false,
        message: `Azure CLI validation failed: ${errorMessage}`,
        recommendations: [
          "Ensure Azure CLI is properly installed",
          "Run 'az login' to authenticate",
          "Check your network connection"
        ]
      };
    }
  }
}

export default MicrosoftAuthClient;
